src/lib/http/http_server.pl

```diff
@@ -0,0 +1,337 @@
+        (
+            read_header_lines(Stream, Lines),
+            [Request|Headers] = Lines,
+            (
+                (phrase(parse_request(Version, Method, Path, Queries), Request), maplist(map_parse_header, Headers, HeadersKV)) -> (
+                        (
+                            member("Content-Length"-ContentLength, HeadersKV) ->
+                                (number_chars(ContentLengthN, ContentLength), get_bytes(Stream, ContentLengthN, Body))
+                                ;true
+                        ),
+                        format("~w ~s\n", [Method, Path]),
+                        (
+                            (http_handler(Method, Pattern, Handler), phrase(path(Pattern), Path)) ->
+                            (
+                                HttpRequest = http_request(HeadersKV, binary(Body), Queries),
+                                HttpResponse = http_response(_, _, _),
+                                (call(Handler, HttpRequest, HttpResponse) ->
+                                    send_response(Stream, HttpResponse)
+                                ;   format(Stream, "HTTP/1.0 500 Internal Server Error\r\n\r\n")
+                                )
+                            )
+                        ;   format(Stream, "HTTP/1.0 404 Not Found\r\n\r\n", [])
+                        )
+                    );(
+                        format(Stream, "HTTP/1.0 400 Bad Request\r\n\r\n", []) % bad format
+                    )
+            ),
+            ! % Remove
+        ), close(Stream)),
+    accept_loop(Socket).
+
+% Helper and recommended predicates
+
+% http_header(Response, HEaderName, Value)
+http_headers(http_request(Headers, _, _), Headers).
+http_headers(http_response(_, _, Headers), Headers).
+
+http_body(http_request(_, binary(ByteBody), _), text(TextBody)) :- chars_utf8bytes(TextBody, ByteBody).
+http_body(http_request(_, Body, _), Body).
+http_body(http_response(_, Body, _), Body).
+
+http_status_code(http_response(StatusCode, _, _), StatusCode).
+
+http_redirect(http_response(307, text("Moved Temporarily"), ["Location"-Uri]), Uri).
+
+http_query(http_request(_, _, Queries), Key, Value) :- member(Key-Value, Queries).
+
+path(Pattern) -->
+    {
+        Pattern =.. Parts,
+        length(Parts, 3),
+        nth0(1, Parts, Pattern0),
+        nth0(2, Parts, PartAtom),
+        (var(PartAtom) -> Part = PartAtom; atom_chars(PartAtom, Part))
+    },
+    path(Pattern0),
+    "/",
+    string_without("/", Part).
+
+path(Pattern) -->
+    {
+        Pattern =.. Parts,
+        Parts = [PartAtom],
+        (var(PartAtom) -> Part = PartAtom; atom_chars(PartAtom, Part))
+    },
+    "/",
+    string_without("/", Part).
+
+path([]) --> [].
+
+send_response(Stream, http_response(StatusCode0, file(Filename), Headers)) :-
+    default(StatusCode0, 200, StatusCode),
+    format(Stream, "HTTP/1.0 ~d\r\n", [StatusCode]),
+    overwrite_header("Connection"-"Close", Headers0, Headers1),
```

Comment:
Sorry, I'm still learning how to read Prolog, but is it a mistake that this is the only use of `Headers0` in this rule?

Reply:
Yes, it's a mistake :rofl: Nice catch!